bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/main/java/com/google/cloud/bigtable/mirroring/hbase1_x/MirroringTable.java

```diff
@@ -0,0 +1,1052 @@
+        };
+
+    this.referenceCounter.holdReferenceUntilCompletion(
+        RequestScheduling.scheduleRequestAndVerificationWithFlowControl(
+            writeOperationInfo.requestResourcesDescription,
+            secondaryResultFutureSupplier,
+            this.mirroringTracer.spanFactory.wrapWriteOperationCallback(writeErrorCallback),
+            flowController,
+            this.mirroringTracer,
+            new Function<Throwable, Void>() {
+              @Override
+              public Void apply(Throwable throwable) {
+                secondaryWriteErrorConsumer.consume(
+                    writeOperationInfo.hBaseOperation, writeOperationInfo.operations, throwable);
+                return null;
+              }
+            }));
+  }
+
+  private void batchSingleWriteOperation(Row operation) throws IOException {
+    Object[] results = new Object[1];
+    try {
+      batchWithSpan(Collections.singletonList(operation), results);
+    } catch (RetriesExhaustedWithDetailsException e) {
+      Throwable exception = e.getCause(0);
+      if (exception instanceof IOException) {
+        throw (IOException) exception;
+      }
+      throw new IOException(exception);
+    } catch (InterruptedException e) {
+      InterruptedIOException interruptedIOException = new InterruptedIOException();
+      interruptedIOException.initCause(e);
+      throw interruptedIOException;
+    }
+  }
+
+  private void batchWithSpan(final List<? extends Row> inputOperations, final Object[] results)
+      throws IOException, InterruptedException {
+    batchWithSpan(inputOperations, results, null);
+  }
+
+  private <R> void batchWithSpan(
+      final List<? extends Row> inputOperations,
+      final Object[] results,
+      @Nullable final Callback<R> callback)
+      throws IOException, InterruptedException {
+    final List<? extends Row> operations = new ArrayList<>(inputOperations);
+    Log.trace("[%s] batch(operations=%s, results)", this.getName(), operations);
+
+    // We store batch results in a internal variable to prevent the user from modifying it when it
+    // might still be used by asynchronous secondary operation.
+    final Object[] internalPrimaryResults = new Object[results.length];
+
+    CallableThrowingIOAndInterruptedException<Void> primaryOperation =
+        new CallableThrowingIOAndInterruptedException<Void>() {
+          @Override
+          public Void call() throws IOException, InterruptedException {
+            if (callback == null) {
+              MirroringTable.this.primaryTable.batch(operations, internalPrimaryResults);
+            } else {
+              MirroringTable.this.primaryTable.batchCallback(
+                  operations, internalPrimaryResults, callback);
+            }
+            return null;
+          }
+        };
+
+    try {
+      if (!this.performWritesConcurrently || !canBatchBePerformedConcurrently(operations)) {
+        sequentialBatch(internalPrimaryResults, operations, primaryOperation);
```

Comment:
please note that operations is an input/output parameter and hbase will modify it during the previous call